src/Microsoft.Extensions.FileProviders.Physical/ConcurrentLruCache.cs

```diff
@@ -0,0 +1,247 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.FileProviders.Physical
+{
+    /// <summary>
+    /// Cache with a fixed size that evicts the least recently used members.
+    /// Thread-safe.
+    /// </summary>
+    internal class ConcurrentLruCache<K, V>
+    {
+        private readonly int _capacity;
+
+        private struct CacheValue
+        {
+            public V Value;
+            public LinkedListNode<K> Node;
+        }
+
+        private readonly Dictionary<K, CacheValue> _cache;
+        private readonly LinkedList<K> _nodeList;
```

Comment:
I hope not to waste your time, but I send you a source code of a `ConcurrentLruCache` with only a concurrent dictionary. I think it minimizes the bottlenecks.